[lldb] refactor highlighting function for image lookup command

junior-jl opened the pull request:
Follow-up to #69422.

This PR puts all the highlighting settings into a single struct for easier handling

Co-authored-by: Talha Tahir <talha.tahir@10xengineers.ai>

taalhaataahir0102 commented:
Hi! @DavidSpickett, made the above changes.
1. struct has been introduced inside stream.h and all the information is being passed using optional struct (`std::optional<Information> pattern_info`) and all the member variables inside the struct are `llvm::StringRef`. Were not able to solve the issue for `llvm::regex`
2. For function search, it was similar to symbol search as it also passes all the information to the same dumping functions. 
3. For test cases, Right now all the test cases are for -s flag (symbol search) i..e, 11 in total. Should we split the current test cases like 4 for symbol search, 4 for function search (-F flag) and 3 for symbol-or-function search (-n flag) or we should add more test cases for each flag?

DavidSpickett commented on lldb/include/lldb/Utility/Stream.h:
```diff
@@ -23,6 +23,16 @@
 
 namespace lldb_private {
 
+struct Information {
```
This should be within the Stream class.

Also the name is too generic, `HighlightSettings` maybe? Choosing a name is hard but something that says what the information is about.

junior-jl replied:
Done.

I was thinking: should we add a default constructor?

DavidSpickett replied:
Looks like you already did this and yes it's a good thing, it means folks can't forget to give all 3 bits of information.

DavidSpickett commented on lldb/source/Commands/CommandObjectTarget.cpp:
```diff
@@ -1609,6 +1612,11 @@ static uint32_t LookupSymbolInModule(CommandInterpreter &interpreter,
+        interpreter.GetDebugger().GetRegexMatchAnsiPrefix();
+    llvm::StringRef ansi_suffix =
+        interpreter.GetDebugger().GetRegexMatchAnsiSuffix();
+    Information info(name, ansi_prefix, ansi_suffix);
```
Construct this closer to the place of use, also inline the `GetRegex...` calls into it like `info(...GetRegexMatch...())`.

junior-jl replied:
I'm unsure if the modification made in last commit addresses this correctly.

DavidSpickett commented on lldb/source/Symbol/Symbol.cpp:
```diff
@@ -262,14 +263,12 @@ void Symbol::GetDescription(Stream *s, lldb::DescriptionLevel level,
     s->PutCString(", mangled=\"");
-    s->PutCStringColorHighlighted(mangled_name.GetStringRef(), pattern,
-                                  ansi_prefix, ansi_suffix);
+    s->PutCStringColorHighlighted(mangled_name.GetStringRef(), pattern_info);
```
The name `pattern_info` for a local variable is fine I think, the name of the type bothers me more.

junior-jl replied:
Since the struct name changed, should we rename this to something like `highlight_settings` or `color_settings`?

DavidSpickett replied:
`highlight_settings`, given that you could use any ANSII sequence. I'm sure you could turn the text upside down if you really wanted to :)

DavidSpickett commented on lldb/source/Utility/Stream.cpp:
```diff
@@ -72,23 +72,21 @@ size_t Stream::PutCString(llvm::StringRef str) {
   }
 
-  llvm::Regex reg_pattern(pattern);
+  llvm::Regex reg_pattern(pattern_info.value().pattern);
```
This isn't wrong but you can instead do:
```
pattern_info->pattern
```
Pointer like operations on optionals accessed the contained value.

DavidSpickett replied:
Which can be cleaner, is my point, `value()` vs. `->`.

junior-jl replied:
Done. Thanks!

DavidSpickett commented:
Drive by comments since I am low on time, also I will be away until early January from today.

A general comment, please keep each PR to one thing. So I'd say making the highlight stuff into a struct is one PR, then adding function highlighting is another. Makes it easier to understand for everyone involved.

On tests for function highlighting, assuming we know that `PutCStringHighlighted` works (which we do), we don't need to repeat single match, multiple match, end of string match, etc. for functions as well. Just check that the parts of the output for functions have some sort of highlight in them. Use the simplest query that will achieve that.

junior-jl committed:
refactor PutCStringColorHighlight | add struct to store highlight settings

junior-jl commented:
Addressed your comments @DavidSpickett. This PR is now focused only on refactoring the colorizing function with the new struct. Adding this feature to function search will be made in the next PR.

DavidSpickett commented on lldb/include/lldb/Utility/Stream.h:
```diff
@@ -260,10 +271,9 @@ class Stream {
-                                  llvm::StringRef suffix = "");
+  void PutCStringColorHighlighted(
+      llvm::StringRef text,
+      std::optional<HighlightSettings> pattern_info = std::nullopt);
```
`pattern_info` relates to the old way of doing it more than this.

How about `settings`? It will be obvious that it's settings for highlighting because it's in `PutCStringColorHighlighted`.

junior-jl replied:
✅

DavidSpickett commented on lldb/source/Commands/CommandObjectTarget.cpp:
```diff
@@ -1618,22 +1621,25 @@ static uint32_t LookupSymbolInModule(CommandInterpreter &interpreter,
-              use_color && name_is_regex ? name : nullptr);
+              use_color && name_is_regex
+                  ? std::optional<Stream::HighlightSettings>{pattern_info}
+                  : std::nullopt);
```
I'm thinking about how to simplify the logic here. While not having to build a struct every single time.

First, you could move the `Stream::HighlightSettings` construction outside the loop ("hoist" some people call that). This means if we match 1 million symbols, we only construct one highlight settings object.

If we know that `use_colour && name_is_regex` is false, we could set `name` in the `HighlightSettings` to `""`. Then when it's passed to `DumpAddress` it will see the pattern is empty and print normally and exit early.

That means you could have:
```
DumpAddress(
              interpreter.GetExecutionContext().GetBestExecutionContextScope(),
              symbol->GetAddressRef(), verbose, all_ranges, strm, pattern_info);
```
Always passing it shouldn't cost that much and removing the ternary removes a potential branch which balances that out a bit.

However, this makes the parameter being `optional` kinda pointless, and I think it should still be optional for every other caller who can just use `nullopt` and not pay the cost of making a `HighlightSettings`.

So in summary: I would move the `HighlightSettings` construction out of the loop. The rest of my idea I think is marginal and risks over complicating things.

junior-jl replied:
Done! ✅

DavidSpickett commented on lldb/source/Utility/Stream.cpp:
```diff
@@ -72,23 +72,20 @@ size_t Stream::PutCString(llvm::StringRef str) {
+void Stream::PutCStringColorHighlighted(
+    llvm::StringRef text, std::optional<HighlightSettings> pattern_info) {
+  // Only apply color formatting when a pattern information is specified.
+  // Otherwise, output the text without color formatting.
```
To link back to my earlier comment, this early exit is fine and should stay.

I think, however, that it's best for callers to pass `nullopt` for the settings if they want to disable highlighting. That's a much more intuitive way to do it, backed up by the type system.

Which is what you're doing, so don't change anything! (just wanted to explain my thinking here)

DavidSpickett commented:
> Do you think the commit message is good enough?

The commit spends a lot of time talking about the previous PR not saying what this PR does. "Follow up to #..." is enough to give context.

This PR puts all the highlighting settings into a single struct for easier handling. That's what the commit message should tell us.

> Also, kind of a git question: is it better to wait this PR conclude to open the next since there will be concurrent modifications in basically the same files?

So you're asking about "stacked PRs" and I myself haven't tried this. LLVM has recently enabled user branches on llvm/llvm-project so I think it is possible using some of the tools out there, but I have not tried it. I would look on the forums for anyone who's tried it, or ask on Discord for advice.

A half way point that you can do, is to have a local branch that has changes for many PRs on it. Then make a branch that only includes the changes meant for the first PR. Push that into review and in the meantime, you can continue to work on the larger branch. When the PR lands, rebase the larger branch onto `main`, and make another sub branch for the next PR, etc.

```
all_changes:
C
B
A
main

<make a new branch with only A changes on main>

all_changes:
C
B            first_pr:
A            A
main         main

<push first_pr for review>
<PR is approved>
<PR landed onto main>
<rebase all_changes>

all_changes:
C
B
main
```

This is what I've been doing, it's got its own problems and if someone wants to see the whole stack, you'll have to push `all_changes` to GitHub also (but for me doing open source work only that's fine). Also I have decent luck with fast reviews, or am doing work where it needs in depth review, so it works out for me.

If you have 10 small changes to make and your reviewer is in the opposite time zone, stacked PRs can be a great benefit as the reviewer can look at all 10 in one session so the feedback is ready for your next working day.

Either method involves some amount rebasing and conflict resolution.

junior-jl committed:
rename pattern_info to settings

junior-jl commented:
Last commit changes:
- Renamed `pattern_info` to `settings` in all of its uses.
- Moved `HighlightSettings` declaration outside `for` loop.
- Edited commit message as suggested.

> So you're asking about "stacked PRs" and I myself haven't tried this. LLVM has recently enabled user branches on llvm/llvm-project so I think it is possible using some of the tools out there, but I have not tried it. I would look on the forums for anyone who's tried it, or ask on Discord for advice.
...

Thanks for the detailed explanation. I'll look better into this.

Question: Would be good/necessary to refactor `LookupSymbolInModule` into some smaller functions? At least separating the printing part?

DavidSpickett commented:
> Question: Would be good/necessary to refactor LookupSymbolInModule into some smaller functions? At least separating the printing part?

Maybe this is me being used to monster functions in lldb, but this one seems fine. At least the volume of code.

What you could do is "early return" (/early continue) in a couple of places to reduce the indentation.

1. You could early return if there are no matches.
2. You could `continue` if `Symbol *symbol` is `nullptr`.

This turns it into mostly straight line code (and from there if you did want to break it into functions, it would be easier).

DavidSpickett commented on lldb/source/Core/Address.cpp:
```diff
@@ -524,8 +525,7 @@ bool Address::Dump(Stream *s, ExecutionContextScope *exe_scope, DumpStyle style,
                     ansi_suffix =
```
Is this unused now?

DavidSpickett commented on lldb/source/Symbol/SymbolContext.cpp:
```diff
@@ -178,8 +170,7 @@ bool SymbolContext::DumpStopContext(Stream *s, ExecutionContextScope *exe_scope,
         ansi_prefix = target_sp->GetDebugger().GetRegexMatchAnsiPrefix();
         ansi_suffix = target_sp->GetDebugger().GetRegexMatchAnsiSuffix();
```
Unused?

DavidSpickett reviewed (commented):
Do a check for unused calls to getregexmatch, I found a few I think.

junior-jl commented:
> Maybe this is me being used to monster functions in lldb, but this one seems fine. At least the volume of code.
What you could do is "early return" (/early continue) in a couple of places to reduce the indentation.
> 1.   You could early return if there are no matches.
> 2.   You could continue if Symbol *symbol is nullptr.

> This turns it into mostly straight line code (and from there if you did want to break it into functions, it would be easier).

I see. But this would need to be in another PR, right? (checking to see if I'm understanding the flow of contribution)

DavidSpickett commented:
> I see. But this would need to be in another PR, right?

Yes.

junior-jl committed:
remove deprecated declarations

junior-jl commented:
> Do a check for unused calls to getregexmatch, I found a few I think.

Done in the last commit. There was the two you mentioned and another one.

DavidSpickett reviewed (approved):
LGTM.

Unless you've got commit access yourself, I can land this for you on Monday.
